File: server/game-engine/src/main/java/com/deepdame/engine/core/logic/MoveValidator.java
package com.deepdame.engine.core.logic;

import com.deepdame.engine.core.model.*;

import java.util.ArrayList;
import java.util.List;

public class MoveValidator {


    // the main function : cchecks if a specific move is allowed
    public boolean isValidMove(Board board, Move move, PieceType player){
        if (move == null) return false;
        List<Move> moveList = getLegalMoves(board, player);

        return moveList.contains(move);
    }


    // this fucntion should scan entire board for a specific player and return all the allowd/legal moves
    public List<Move> getLegalMoves(Board board, PieceType player){
        List<Move> jumps = new ArrayList<>();
        List<Move> slides = new ArrayList<>();


        for(int row = 0; row < board.SIZE; row++){
            for(int col = 0; col < board.SIZE; col++){
                Position pos = new Position(row, col);
                Piece piece = board.getPiece(pos);
                if(piece != null && piece.type().equals(player)){
                    findMovesForPiece(board, pos, piece, jumps, slides);
                }
            }

        }


        // foreced jump rule
        if (!jumps.isEmpty()) {
            return jumps;
        }
        return slides;
    }

    private int[][] getDirections(Piece piece){
        if (piece.isKing()) {
            return new int[][]{{-1, -1}, {-1, 1}, {1, -1}, {1, 1}};
        }
        if (piece.type() == PieceType.WHITE) {
            return new int[][]{{1, -1}, {1, 1}};
        } else {
            return new int[][]{{-1, -1}, {-1, 1}};
        }
    }

    private void findMovesForPiece(Board board, Position pos, Piece piece, List<Move> jumps, List<Move> slides){
        int[][] directions = getDirections(piece);
        for(int[] direction : directions){
            int dRow = direction[0];
            int dCol = direction[1];

            Position next = new Position(pos.row() + dRow, pos.col() + dCol);
            Position landing = new Position(pos.row() + (2 * dRow), pos.col() + (2 * dCol));
            Piece middlePiece = board.getPiece(next);
            Piece landingPiece = board.getPiece(landing);

            if (middlePiece != null && !middlePiece.type().equals(piece.type())) {

                if (board.isValidBounds(landing) && landingPiece == null) {
                    jumps.add(new Move(pos, landing));
                }
            }

            if (board.isValidBounds(next) && board.getPiece(next) == null) {
                slides.add(new Move(pos, next));
            }
        }
    }

}

File: client/deepdame/lib/game-engine/logic/move_validator.dart
import '../model/board.dart';
import '../model/move.dart';
import '../model/piece.dart';
import '../model/piece_type.dart';
import '../model/position.dart';

class MoveValidator {
  
  bool isValidMove(Board board, Move? move, PieceType player) {
    if (move == null) return false;
    
    List<Move> moveList = getLegalMoves(board, player);
    return moveList.contains(move);
  }

  List<Move> getLegalMoves(Board board, PieceType player) {
    List<Move> jumps = [];
    List<Move> slides = [];

    for (int row = 0; row < Board.size; row++) {
      for (int col = 0; col < Board.size; col++) {
        Position pos = Position(row, col);
        Piece? piece = board.getPiece(pos);
        
        if (piece != null && piece.type == player) {
          _findMovesForPiece(board, pos, piece, jumps, slides);
        }
      }
    }

    if (jumps.isNotEmpty) {
      return jumps;
    }
    return slides;
  }

  List<List<int>> _getDirections(Piece piece) {
    if (piece.isKing) {
      return [
        [-1, -1], [-1, 1], 
        [1, -1], [1, 1]
      ];
    }
    if (piece.type == PieceType.white) {
      return [
        [1, -1], [1, 1]
      ];
    } else {
      return [
        [-1, -1], [-1, 1]
      ];
    }
  }

  void _findMovesForPiece(
    Board board, 
    Position pos, 
    Piece piece, 
    List<Move> jumps, 
    List<Move> slides
  ) {
    List<List<int>> directions = _getDirections(piece);

    for (List<int> direction in directions) {
      int dRow = direction[0];
      int dCol = direction[1];

      Position next = Position(pos.row + dRow, pos.col + dCol);
      Position landing = Position(pos.row + (2 * dRow), pos.col + (2 * dCol));

      Piece? middlePiece = board.getPiece(next);
      Piece? landingPiece = board.getPiece(landing);

      if (middlePiece != null && middlePiece.type != piece.type) {
        if (board.isValidBounds(landing) && landingPiece == null) {
          jumps.add(Move(pos, landing));
        }
      }

      if (board.isValidBounds(next) && board.getPiece(next) == null) {
        slides.add(Move(pos, next));
      }
    }
  }
}